python/taichi/misc/gui.py

```diff
@@ -139,24 +139,57 @@ def show(self, file=None):
             self.core.screenshot(file)
         self.clear(self.background_color)
 
+    class EventFilter:
+        def __init__(self, *filter):
+            self.types = set()
+            self.keys = set()
+            self.combs = set()
+            for ent in filter:
+                if isinstance(ent, (tuple, list)):
+                    type, key = ent
+                    self.combs.add((type, key))
+                elif ent in [GUI.PRESS, GUI.RELEASE]:
+                    self.types.add(ent)
+                else:
+                    self.keys.add(ent)
+
+        def match(self, e):
+            if (e.type, e.key) in self.combs:
+                return True
+            if not self.keys or e.key in self.keys:
+                if not self.combs or e.key in self.combs:
+                    return True
+            return False
+
     def has_key_event(self):
         return self.core.has_key_event()
 
     def get_event(self, *filter):
+        for e in self.get_events(*filter):
+            self.event = e
+            return True
+        return False
+
+    def get_events(self, *filter):
+        if not filter:
+            filter = None
+        elif not isinstance(filter[0], GUI.EventFilter):
+            filter = GUI.EventFilter(*filter)
+
         while True:
             if not self.has_key_event():
-                return False
-            self.event = self.get_key_event()
-            if not len(filter) or self.event.type in filter:
                 break
-        return True
+            e = self.get_key_event()
+            if filter is None or filter.match(e):
+                yield e
 
     def get_key_event(self):
         self.core.wait_key_event()
         e = GUI.Event()
         e.key = self.core.get_key_event_head_key()
         e.type = self.core.get_key_event_head_type()
         e.pos = self.core.get_key_event_head_pos()
+        e.pos = (e.pos[0], e.pos[1])
```

Comment:
nit: I think it's fine just to write `tuple(self.core.get_key_event_head_pos())`, without worrying  too much about the returned size. Or maybe `x, y = self.core.get_key_event_head_pos(); e.pos = (x, y)`?

Reply:
Ack